Scoreboard Creation

SamuSonoIo opened the pull request:
Hello! 👋

I noticed that many people, including myself, struggle with creating scoreboards for Minecraft plugins. So, I decided to add this feature to Twilight to make it easier for everyone!

The process of creating a scoreboard is fully explained in the README.md file, and I've made sure that everything is easy to use and well-explained (at least, I believe so).

With my feature, you can create either a Static or a Dynamic scoreboard, or even combine the two. For example, you can create a static scoreboard and then update only specific lines (though, I personally don't see much use for this, but it's possible!).

If there is something in the code that maybe isn't clean, or need a refactoring, let me know and i'll fix it up as soon as i can.

Thank you!

![image](https://github.com/user-attachments/assets/d1caa840-3f99-455c-ae15-bb0a45a1980f)

Dawsson commented:
I love you

SamuSonoIo commented:
I love you too dawson

Dawsson commented:
Is this the start of our relationship?

SamuSonoIo commented:
Yes😍

Dawsson commented:
yea gl getting anyone to merge this within 2024 💀

SamuSonoIo commented:
I've found a small error while testing today,if anyone looks at this and wants to merge, please wait until my next message as I'm fixing this issue of tomorrow

ShreyasAyyengar commented:
> I've found a small error while testing today,if anyone looks at this and wants to merge, please wait until my next message as I'm fixing this issue tomorrow

If that's the case, make use of the 'Draft Pull Requests' feature to ensure that merging cannot happen if you're not ready to make changes.

<img width="768" alt="Screenshot 2024-12-10 at 6 39 47 pm" src="https://github.com/user-attachments/assets/bfcff01f-2a03-4e91-bf47-4436300d0d46">

stephendotgg commented:
@SamuSonoIo Please ping once ready to be merged again

stephendotgg reviewed (approved):
Thanks for this! I like the syntax. And thanks for adding to the readme. I'll make a few changes to that for my OCD before I publish. Not going to be anal about the code because no need.

@joshbker Any adjustments wanted?

SamuSonoIo commented:
Awesome!

joshbker commented:
> Thanks for this! I like the syntax. And thanks for adding to the readme. I'll make a few changes to that for my OCD before I publish. Not going to be anal about the code because no need.
> 
> @joshbker Any adjustments wanted?

I also just have some nitpicks around naming, other than that it all looks great. I'm happy to change any nitpicks myself though.

I do have some questions though @SamuSonoIo:

1. Should we have an example for having two different boards and switching? What is the best way to do this? Two board instances and assigning them, or a better way of controlling the lines list?

2. Should we change some of the method naming? It doesn't feel very "Kotliny", we lost setters and getters, so the "set" naming feels weird.

3. Also on naming, is "title" more apt than "name" for the scoreboard title? I've always heard it referred to as a title not a name.

4. How do I unassign a scoreboard from a player? I assume I have to just do `player.scoreboard = Bukkit.getScoreboardManager().getMainScoreboard()` or whatever it is, but would be nice to have a util method as a part of this.

5. How would someone manage multiple objectives in this scoreboard? We're forcefully setting the player's scoreboard, what if someone wants an objective for showing something in the slot under names, etc?

joshbker commented:
> Hey @joshbker Sure! Happy it looks great, to answer your question:
> 
>     1. For the example part, sure, to have two different boards there is just the need of creating 2 instances of TwilightScoreboard and set that when needed
> 
>     2. I know sadly, but i didn't really know how to name those methods, like "set" let you set a specific line, so, i haven't really found an alternative yet xD
> 
>     3. Yeah sure, i'll rename it to title then. sure sounds better
> 
>     4. Exactly that is the method, but we can easily add a method to do this instead of getting the scoreboard manager.
> 
>     5. I'm not sure i understood your question, but i guess you mean like what if i want only to change one score? if so there is the .set method that lets you change a specific line so you don't need to remake the scoreboard.
> 
>     6. I agree with you, i believe we can use minimessage and instead of using components you just use string with minimessages tag to customize the message.
> 
> 
> Let me know if this answer your questions.

1. Cool.

2. Let's change `setName` to `title`.

3. ^

4. Sounds good, let's add that method.

5. For example, if you wanted to display something like health under player names, that is an objective you register to the scoreboard - we can't do that easily right now whilst having a TwilightScoreboard assigned to a player, unless I misunderstand the Scoreboard/Teams API.

6. Worth a try, I really dislike how repetitive it is, would be nice to give more options.

SamuSonoIo commented:
To answer question 5:
Yes, right now i haven't really added a method to do it easily.
I'll probably work on it soon then; for now to add a new line you would need to use the setAll or updateLines method again and re set it to the player i believe, but will find a solution.

For question num 6:
Sure, shouldn't be to hard

joshbker commented:
> To answer question 5: Yes, right now i haven't really added a method to do it easily. I'll probably work on it soon then; for now to add a new line you would need to use the setAll or updateLines method again and re set it to the player i believe, but will find a solution.
> 
> For question num 6: Sure, shouldn't be to hard

You're still misunderstanding 5.

I'm talking about adding another Objective in another DisplaySlot, such as BELOW_NAME. This is bound to the org.bukkit.Scoreboard, no?

If someone wants to have for example, an Objective for player health display BELOW_NAME, currently they would have to choose between that and a TwilightScoreboard SIDEBAR.

Please correct me if I'm wrong, but that's how I understand it currently.

joshbker commented:
Also @SamuSonoIo, have you tested the example code where it references `player.name` with more than 1 player online?

I'm not sure if I'm being stupid but I can't see how these scoreboards are per-player.

If I join then the scoreboard would say my name, then if you join after it would say your name for both of us I think.

SamuSonoIo commented:
Oh, I understand now.
For now, I've only implemented the SIDEBAR display and not the BELOW_NAME. 
It is still bound to the org.bukkit.Scoreboard, as you said. When I started this project,
I didn't consider that use case, but I can add support for it.

SamuSonoIo commented:
@joshbker 
I didn't test this with multiple player since i didn't have the chance.
But i guess if someone create the scoreboard instance every time someone joins the game every scoreboard should have it's own runnable so it's own name and it's own things.

joshbker commented:
> Oh, I understand now. For now, I've only implemented the SIDEBAR display and not the BELOW_NAME. It is still bound to the org.bukkit.Scoreboard, as you said. When I started this project, I didn't consider that use case, but I can add support for it.

Alright, yes I think as a library we really need to support those kind of things.

> @joshbker I didn't test this with multiple player since i didn't have the chance. But i guess if someone create the scoreboard instance every time someone joins the game every scoreboard should have it's own runnable so it's own name and it's own things.

You don't create a new scoreboard instance per player in the README:
![image](https://github.com/user-attachments/assets/98f80d11-78c0-43a5-b45d-ebbd48ea8907)

SamuSonoIo commented:
Oh my bad then, i made an error in the readme.
Sorry

SamuSonoIo commented:
Will finish to fix everything tomorrow probably,
Thanks for the feedback <3
Anything else that could be useful in this?

joshbker commented:
> Will finish to fix everything tomorrow probably, Thanks for the feedback <3 Anything else that could be useful in this?

I can't think of anything more at the moment, think if you/we address all of the thinggs we've spoken about so far then we'll be in a great position with this.

joshbker commented on src/main/kotlin/gg/flyte/twilight/scoreboard/TwilightScoreboard.kt:
```diff
@@ -0,0 +1,156 @@
+
+class TwilightScoreboard(private val plugin: JavaPlugin) {
+    private val scoreboard: Scoreboard = Bukkit.getScoreboardManager().newScoreboard
+    private var objective = scoreboard.registerNewObjective("sidebar", "dummy")
```
This objective name should probably be "twilight-sidebar" or something so we can easily obtain all Twilight related ones.

joshbker commented on src/main/kotlin/gg/flyte/twilight/scoreboard/TwilightScoreboard.kt:
```diff
@@ -0,0 +1,156 @@
+     * @param player the player you want to remove the scoreboard to
+     * Assign the scoreboard to a player (Use a for Loop if you want to assign this to multiple players)
+     */
+    fun removeTo(player: Player) {
```
"removeTo" doesn't make grammatical sense, maybe just call this "remove", or you could say "removeFrom"

SamuSonoIo commented:
Hi @joshbker 
What do you think about this for the below name?
(Don't mind the semicolon there i was testing around)
![image](https://github.com/user-attachments/assets/ef39ac93-f683-469d-80c5-ef4203ad6355)
![image](https://github.com/user-attachments/assets/77ae8c16-3aed-429b-b835-9d26ebf69732)

SamuSonoIo commented:
![image](https://github.com/user-attachments/assets/ebb9f9a9-ecf8-4f01-8490-91168f403865)
Hi!
Also works with multiple player with just one instance.

SamuSonoIo commented:
I'm also working on some TabList methods since people love to use them together.
![image](https://github.com/user-attachments/assets/6eb18ce5-952b-4378-8054-5e79cf0a3632)
I tried almost everything, there are 2 small issues i still need to fix (which i will fix after christmas) then i think it's good to go!

joshbker commented:
> I'm also working on some TabList methods since people love to use them together. ![image](https://private-user-images.githubusercontent.com/176231328/398423816-6eb18ce5-952b-4378-8054-5e79cf0a3632.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU2MDY1NjUsIm5iZiI6MTczNTYwNjI2NSwicGF0aCI6Ii8xNzYyMzEzMjgvMzk4NDIzODE2LTZlYjE4Y2U1LTk1MmItNDM3OC04MDU0LTVlNzljZjBhMzYzMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjMxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIzMVQwMDUxMDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05ZWQ5MThlNTE3NzYyOWZkY2UxMDI5YmZiZWNhY2E3OWJiNjJjYjkwODMxOGZkNmEyNmE0OGY0MTJkMTU1ZWZjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Dv49RZGHBInDVsQbtIIMcAqJbAN2yPw0CxawY-WVwzA) I tried almost everything, there are 2 small issues i still need to fix (which i will fix after christmas) then i think it's good to go!

Awesome, thank you for all your hard work on this!

SamuSonoIo commented:
Hey!  
I've been working on this for a while, and I think it's finally good to go.  
I updated the ReadMe and added some examples too.  

I basically remade everything, now the entire scoreboard system is in one class, and it only has the methods that are actually needed, no unnecessary ones, so everything is nice and clean.  
I took some inspiration from FastBoard since people seem to really like that API...

That's about it!

ShreyasAyyengar commented on src/main/kotlin/gg/flyte/twilight/string/String.kt:
```diff
@@ -212,4 +216,19 @@ fun String.formatCase(case: Case): String = CASE_DELIMITER_REGEX.split(this)
  */
 fun String.capitalizeFirstLetter(): String = replaceFirstChar {
     if (it.isLowerCase()) it.titlecase() else it.toString()
+}
+
+/**
+ * Minimessage builder
+ * Some useful methods to use around the code if needed.
+ */
+val miniMessage = MiniMessage.builder().build()
+
+fun String.toComponent(): TextComponent = Component.text(this)
+
+fun String.toMini(): Component = miniMessage.deserialize(this)
+
+fun String.toMini(vararg placeholders: Component): Component {
+    val components = placeholders.mapIndexed { i, it -> Placeholder.component(i.toString(), it) }.toTypedArray()
+    return miniMessage.deserialize(this, *components)
```
I personally think that this is something that should be left up to the implementing developer, rather than something Twilight should provide directly. I agree it's useful to offer such utility methods, I have a feeling that it's introducing unnecessary coupling or assumptions about how components are built or serialized. 

Many (myself included) choose to build their components with difference tag resolvers or placeholders, so a future PR could focus on giving them these utility methods with the option to configure their placeholders at the Twilight init stage. 

As for this PR directly, I think it's unnecessary.

SamuSonoIo replied:
Hi!
Thanks for the feedback, the reason i actually implemented this is because when creating a Scoreboard it's  annoying to always do Component.text, so i just sticked to using minimessage tags, couldn't think of other options, if you have any other ideas, i will be happy to hear and implement them!

ShreyasAyyengar replied:
I would've recommended overloading the methods with a String parameter and then parsing `Component.text(String)` to your internals, but even that's a convoluted solution because then it must only be a plain text String. (e.g. you can't parse `<red>My epic scoreboard`)

The best approach is to supply `updateSidebarTitle`, `updateSidebarLines`, `updateTabList`, `belowName`, `updateBelowNameScore`, and `prefix` and `suffix` with parameters as `Component` rather than `String`. 

As I see it, this PR aims to make Scoreboard/Tablist modification easier. However, by forcing users to use a string, you're assuming they will 100% use a MiniMessage-compatible String to deserialize into a component, which will not always be the case.

SamuSonoIo replied:
Okay, sounds good. 
@joshbker  Hey, what do you think about this?

joshbker replied:
I like Shreyas' take I think @SamuSonoIo. Overall I am happy with this PR though

ShreyasAyyengar commented on README.md:
```diff
@@ -180,7 +180,7 @@ event<ChatClickEvent> {
     when (data[1]) {
         "warps" -> GUIManager.openWarps(player)
-         ...
+            ...
```
```suggestion
         // {...}
```

SamuSonoIo commented:
![image](https://github.com/user-attachments/assets/98300cbd-0ff1-4bec-ab81-369c230a4838)
Hi,
Done, everything is now using regular components, going to update the README so it uses components and not regular string.
